extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java

```diff
@@ -148,15 +161,104 @@ void build(CombinedIndexBuildItem index,
         MetamodelInfo<EntityModel<EntityField>> modelInfo = modelEnhancer.getModelInfo();
         if (modelInfo.hasEntities()) {
             PanacheFieldAccessEnhancer panacheFieldAccessEnhancer = new PanacheFieldAccessEnhancer(modelInfo);
-            for (ClassInfo classInfo : index.getIndex().getKnownClasses()) {
-                String className = classInfo.name().toString();
-                if (!modelClasses.contains(className)) {
-                    transformers.produce(new BytecodeTransformerBuildItem(className, panacheFieldAccessEnhancer));
+            QuarkusClassLoader tccl = (QuarkusClassLoader) Thread.currentThread().getContextClassLoader();
+            List<ClassPathElement> archives = tccl.getElementsWithResource(META_INF_PANACHE_ARCHIVE_MARKER);
+            for (ClassPathElement i : archives) {
+                for (String res : i.getProvidedResources()) {
+                    if (res.endsWith(".class")) {
+                        String cn = res.replace("/", ".").substring(0, res.length() - 6);
+                        if (!modelClasses.contains(cn)) {
+                            if (requiresTransformation(res, i, modelClassNamesInternal)) {
+                                transformers.produce(new BytecodeTransformerBuildItem(cn, panacheFieldAccessEnhancer));
+                            }
+                        }
+                    }
                 }
             }
         }
     }
 
+    static final int CONSTANT_UTF8_TAG = 1;
+    static final int CONSTANT_INTEGER_TAG = 3;
+    static final int CONSTANT_FLOAT_TAG = 4;
+    static final int CONSTANT_LONG_TAG = 5;
+    static final int CONSTANT_DOUBLE_TAG = 6;
+    static final int CONSTANT_CLASS_TAG = 7;
+    static final int CONSTANT_STRING_TAG = 8;
+    static final int CONSTANT_FIELDREF_TAG = 9;
+    static final int CONSTANT_METHODREF_TAG = 10;
+    static final int CONSTANT_INTERFACE_METHODREF_TAG = 11;
+    static final int CONSTANT_NAME_AND_TYPE_TAG = 12;
+    static final int CONSTANT_METHOD_HANDLE_TAG = 15;
+    static final int CONSTANT_METHOD_TYPE_TAG = 16;
+    static final int CONSTANT_DYNAMIC_TAG = 17;
+    static final int CONSTANT_INVOKE_DYNAMIC_TAG = 18;
+    static final int CONSTANT_MODULE_TAG = 19;
+    static final int CONSTANT_PACKAGE_TAG = 20;
+
+    private boolean requiresTransformation(String res, ClassPathElement i, Set<String> modelInfo) {
```

Comment:
I think it would be better to move this in a core class like ConstantPoolScanner and make it take a ConstantPoolPredicate that can match, no?
I want to reuse this for https://github.com/quarkusio/quarkus/issues/814 :)
Also, I wonder how we can make sure we update this scanner at every new JVM spec that defines a new constant pool type.

Reply:
Every time there is a new JVM spec we need to update ASM, ByteBuddy etc, this is just one more thing